File: src/px4_bridge_cpp/src/cmd_vel_to_pos_cmd.cpp
#include <array>
#include <cmath>
#include <memory>
#include <optional>
#include <string>
#include <utility>

#include "geometry_msgs/msg/twist.hpp"
#include "px4_msgs/msg/vehicle_local_position.hpp"
#include "quadrotor_msgs/msg/position_command.hpp"
#include "rclcpp/create_timer.hpp"
#include "rclcpp/rclcpp.hpp"

namespace
{

constexpr char kLocalPositionTopic[] = "/fmu/out/vehicle_local_position_v1";

}  // namespace

class CmdVelToPosCmdBridge : public rclcpp::Node
{
public:
  CmdVelToPosCmdBridge()
  : Node("cmd_vel_to_pos_cmd"),
    current_position_{0.0, 0.0, 0.0}
  {
    cmd_vel_topic_ = declare_parameter<std::string>("cmd_vel_topic", "/cmd_vel");
    pos_cmd_topic_ = declare_parameter<std::string>("pos_cmd_topic", "/drone_0_planning/pos_cmd");
    frame_id_ = declare_parameter<std::string>("frame_id", "map");
    update_rate_hz_ = std::max(declare_parameter<double>("update_rate_hz", 30.0), 1.0);
    cmd_vel_timeout_sec_ = std::max(declare_parameter<double>("cmd_vel_timeout_sec", 0.5), 0.0);

    auto qos_profile_sub = rclcpp::QoS(rclcpp::KeepLast(1)).best_effort().durability_volatile();

    cmd_vel_sub_ = create_subscription<geometry_msgs::msg::Twist>(
      cmd_vel_topic_,
      rclcpp::QoS(10),
      std::bind(&CmdVelToPosCmdBridge::cmdVelCallback, this, std::placeholders::_1));

    local_position_sub_ = create_subscription<px4_msgs::msg::VehicleLocalPosition>(
      kLocalPositionTopic,
      qos_profile_sub,
      std::bind(&CmdVelToPosCmdBridge::vehicleLocalPositionCallback, this, std::placeholders::_1));

    pos_cmd_pub_ = create_publisher<quadrotor_msgs::msg::PositionCommand>(pos_cmd_topic_, 10);

    const auto period = rclcpp::Duration::from_seconds(1.0 / update_rate_hz_);
    timer_ = rclcpp::create_timer(
      this,
      get_clock(),
      period,
      std::bind(&CmdVelToPosCmdBridge::timerCallback, this));

    RCLCPP_INFO(
      get_logger(),
      "Planar cmd_vel bridge ready: %s -> %s using %s",
      cmd_vel_topic_.c_str(),
      pos_cmd_topic_.c_str(),
      kLocalPositionTopic);
  }

private:
  static double convertHeadingToRosYaw(double heading)
  {
    return std::atan2(std::cos(heading), std::sin(heading));
  }

  static double wrapAngle(double angle)
  {
    return std::atan2(std::sin(angle), std::cos(angle));
  }

  void vehicleLocalPositionCallback(const px4_msgs::msg::VehicleLocalPosition::SharedPtr msg)
  {
    if (!std::isfinite(msg->x) || !std::isfinite(msg->y) ||
      !std::isfinite(msg->z) || !std::isfinite(msg->heading))
    {
      return;
    }

    current_position_ = {
      static_cast<double>(msg->y),
      static_cast<double>(msg->x),
      static_cast<double>(-msg->z),
    };
    current_yaw_ = convertHeadingToRosYaw(static_cast<double>(msg->heading));
    has_pose_ = true;
  }

  void cmdVelCallback(const geometry_msgs::msg::Twist::SharedPtr msg)
  {
    latest_cmd_linear_x_ = static_cast<double>(msg->linear.x);
    latest_cmd_linear_y_ = static_cast<double>(msg->linear.y);
    latest_cmd_yaw_rate_ = static_cast<double>(msg->angular.z);
    last_cmd_vel_time_ = get_clock()->now();
    has_cmd_vel_ = true;
  }

  bool hasRecentCmdVel()
  {
    if (!has_cmd_vel_) {
      return false;
    }
    if (cmd_vel_timeout_sec_ <= 0.0) {
      return true;
    }
    return (get_clock()->now() - last_cmd_vel_time_).seconds() <= cmd_vel_timeout_sec_;
  }

  std::pair<double, double> bodyToWorldPlanarVelocity(
    double forward_velocity,
    double left_velocity) const
  {
    const double cos_yaw = std::cos(current_yaw_);
    const double sin_yaw = std::sin(current_yaw_);
    const double world_x = forward_velocity * cos_yaw - left_velocity * sin_yaw;
    const double world_y = forward_velocity * sin_yaw + left_velocity * cos_yaw;
    return {world_x, world_y};
  }

  void publishPosCmd(
    const std::array<double, 3> & target_position,
    const std::array<double, 3> & target_velocity,
    double target_yaw,
    double target_yaw_rate)
  {
    quadrotor_msgs::msg::PositionCommand msg;
    msg.header.stamp = get_clock()->now();
    msg.header.frame_id = frame_id_;
    msg.position.x = target_position[0];
    msg.position.y = target_position[1];
    msg.position.z = target_position[2];
    msg.velocity.x = target_velocity[0];
    msg.velocity.y = target_velocity[1];
    msg.velocity.z = target_velocity[2];
    msg.acceleration.x = 0.0;
    msg.acceleration.y = 0.0;
    msg.acceleration.z = 0.0;
    msg.yaw = target_yaw;
    msg.yaw_dot = target_yaw_rate;
    pos_cmd_pub_->publish(msg);
  }

  void timerCallback()
  {
    if (!has_pose_) {
      if (!pose_wait_log_sent_) {
        pose_wait_log_sent_ = true;
        RCLCPP_WARN(get_logger(), "Waiting for PX4 local position before forwarding cmd_vel");
      }
      return;
    }

    if (pose_wait_log_sent_) {
      pose_wait_log_sent_ = false;
      RCLCPP_INFO(get_logger(), "Received PX4 pose, cmd_vel bridge is active");
    }

    std::array<double, 3> target_position = current_position_;
    std::array<double, 3> target_velocity = {0.0, 0.0, 0.0};
    double target_yaw = current_yaw_;
    double target_yaw_rate = 0.0;

    if (!hasRecentCmdVel()) {
      hold_z_ = current_position_[2];
    } else {
      if (!hold_z_.has_value()) {
        hold_z_ = current_position_[2];
      }

      const double dt = 1.0 / update_rate_hz_;
      const auto [target_vel_x, target_vel_y] = bodyToWorldPlanarVelocity(
        latest_cmd_linear_x_,
        latest_cmd_linear_y_);

      target_position[0] += target_vel_x * dt;
      target_position[1] += target_vel_y * dt;
      target_velocity[0] = target_vel_x;
      target_velocity[1] = target_vel_y;
      target_yaw = wrapAngle(current_yaw_ + latest_cmd_yaw_rate_ * dt);
      target_yaw_rate = latest_cmd_yaw_rate_;
    }

    target_position[2] = hold_z_.value_or(current_position_[2]);
    publishPosCmd(target_position, target_velocity, target_yaw, target_yaw_rate);
  }

  std::string cmd_vel_topic_;
  std::string pos_cmd_topic_;
  std::string frame_id_;

  double update_rate_hz_{30.0};
  double cmd_vel_timeout_sec_{0.5};

  rclcpp::Subscription<geometry_msgs::msg::Twist>::SharedPtr cmd_vel_sub_;
  rclcpp::Subscription<px4_msgs::msg::VehicleLocalPosition>::SharedPtr local_position_sub_;
  rclcpp::Publisher<quadrotor_msgs::msg::PositionCommand>::SharedPtr pos_cmd_pub_;
  rclcpp::TimerBase::SharedPtr timer_;

  std::array<double, 3> current_position_;
  double current_yaw_{0.0};
  double latest_cmd_linear_x_{0.0};
  double latest_cmd_linear_y_{0.0};
  double latest_cmd_yaw_rate_{0.0};
  rclcpp::Time last_cmd_vel_time_{0, 0, RCL_ROS_TIME};
  std::optional<double> hold_z_;
  bool has_pose_{false};
  bool has_cmd_vel_{false};
  bool pose_wait_log_sent_{false};
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CmdVelToPosCmdBridge>());
  rclcpp::shutdown();
  return 0;
}

File: src/px4_ego_py/px4_ego_py/cmd_vel_to_pos_cmd.py
#!/usr/bin/env python3

import math

import rclpy
from geometry_msgs.msg import Twist
from px4_msgs.msg import VehicleLocalPosition, VehicleOdometry
from quadrotor_msgs.msg import PositionCommand
from rclpy.node import Node
from rclpy.qos import (
    QoSDurabilityPolicy,
    QoSHistoryPolicy,
    QoSProfile,
    QoSReliabilityPolicy,
)


class CmdVelToPosCmdBridge(Node):
    """Bridge Nav2 cmd_vel to quadrotor PositionCommand for offboard_control_test."""

    def __init__(self):
        super().__init__('cmd_vel_to_pos_cmd')

        self.declare_parameter('cmd_vel_topic', '/cmd_vel')
        self.declare_parameter('pos_cmd_topic', '/drone_0_planning/pos_cmd')
        self.declare_parameter('frame_id', 'map')
        self.declare_parameter('update_rate_hz', 50.0)
        self.declare_parameter('cmd_vel_timeout_sec', 0.5)
        self.declare_parameter('position_lookahead_sec', 0.50)
        self.declare_parameter('yaw_lookahead_sec', 0.35)

        self.cmd_vel_topic = str(self.get_parameter('cmd_vel_topic').value)
        self.pos_cmd_topic = str(self.get_parameter('pos_cmd_topic').value)
        self.frame_id = str(self.get_parameter('frame_id').value)
        self.update_rate_hz = max(float(self.get_parameter('update_rate_hz').value), 1.0)
        self.cmd_vel_timeout_sec = max(float(self.get_parameter('cmd_vel_timeout_sec').value), 0.0)
        self.position_lookahead_sec = max(
            float(self.get_parameter('position_lookahead_sec').value),
            0.0,
        )
        self.yaw_lookahead_sec = max(float(self.get_parameter('yaw_lookahead_sec').value), 0.0)

        qos_profile_sub = QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            durability=QoSDurabilityPolicy.VOLATILE,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1,
        )

        self.cmd_vel_sub = self.create_subscription(Twist, self.cmd_vel_topic, self.cmd_vel_callback, 10)
        self.vehicle_local_position_sub = self.create_subscription(
            VehicleLocalPosition,
            '/fmu/out/vehicle_local_position_v1',
            self.vehicle_local_position_callback,
            qos_profile_sub,
        )
        self.vehicle_visual_odom_sub = self.create_subscription(
            VehicleOdometry,
            '/fmu/in/vehicle_visual_odometry',
            self.vehicle_visual_odom_callback,
            qos_profile_sub,
        )

        self.pos_cmd_pub = self.create_publisher(PositionCommand, self.pos_cmd_topic, 10)

        self.current_position = None
        self.current_yaw = 0.0
        self.latest_cmd_linear_x = 0.0
        self.latest_cmd_linear_y = 0.0
        self.latest_cmd_linear_z = 0.0
        self.latest_cmd_yaw_rate = 0.0
        self.last_cmd_vel_time = None
        self.pose_source = 'none'
        self.pose_wait_log_sent = False

        timer_period = 1.0 / self.update_rate_hz
        self.timer = self.create_timer(timer_period, self.timer_callback)

        self.get_logger().info(
            f'cmd_vel bridge ready: {self.cmd_vel_topic} -> {self.pos_cmd_topic}'
        )

    def quaternion_to_yaw(self, w, x, y, z):
        siny_cosp = 2.0 * (w * z + x * y)
        cosy_cosp = 1.0 - 2.0 * (y * y + z * z)
        return math.atan2(siny_cosp, cosy_cosp)

    def convert_yaw_between_fmu_and_ros(self, yaw_value):
        return math.atan2(math.cos(yaw_value), math.sin(yaw_value))

    def wrap_angle(self, angle):
        return math.atan2(math.sin(angle), math.cos(angle))

    def update_current_pose(self, ros_x, ros_y, ros_z, ros_yaw, source):
        self.current_position = [float(ros_x), float(ros_y), float(ros_z)]
        self.current_yaw = float(ros_yaw)
        if self.pose_source != source:
            self.pose_source = source
            self.get_logger().info(f'Pose source switched to {source}')

    def vehicle_local_position_callback(self, msg):
        self.update_current_pose(
            ros_x=msg.y,
            ros_y=msg.x,
            ros_z=-msg.z,
            ros_yaw=self.convert_yaw_between_fmu_and_ros(msg.heading),
            source='vehicle_local_position',
        )

    def vehicle_visual_odom_callback(self, msg):
        q_att = msg.q
        yaw_ned = self.quaternion_to_yaw(q_att[0], q_att[1], q_att[2], q_att[3])
        self.update_current_pose(
            ros_x=msg.position[1],
            ros_y=msg.position[0],
            ros_z=-msg.position[2],
            ros_yaw=self.convert_yaw_between_fmu_and_ros(yaw_ned),
            source='vehicle_visual_odometry',
        )

    def cmd_vel_callback(self, msg):
        self.latest_cmd_linear_x = float(msg.linear.x)
        self.latest_cmd_linear_y = float(msg.linear.y)
        self.latest_cmd_linear_z = float(msg.linear.z)
        self.latest_cmd_yaw_rate = float(msg.angular.z)
        self.last_cmd_vel_time = self.get_clock().now()

    def has_recent_cmd_vel(self):
        if self.last_cmd_vel_time is None:
            return False
        if self.cmd_vel_timeout_sec <= 0.0:
            return True

        age_sec = (self.get_clock().now().nanoseconds - self.last_cmd_vel_time.nanoseconds) / 1e9
        return age_sec <= self.cmd_vel_timeout_sec

    def body_to_world_planar_velocity(self, forward_velocity, left_velocity):
        cos_yaw = math.cos(self.current_yaw)
        sin_yaw = math.sin(self.current_yaw)
        world_x = forward_velocity * cos_yaw - left_velocity * sin_yaw
        world_y = forward_velocity * sin_yaw + left_velocity * cos_yaw
        return world_x, world_y

    def publish_pos_cmd(self, target_position, target_velocity, target_yaw, target_yaw_rate):
        msg = PositionCommand()
        msg.header.stamp = self.get_clock().now().to_msg()
        msg.header.frame_id = self.frame_id
        msg.position.x = float(target_position[0])
        msg.position.y = float(target_position[1])
        msg.position.z = float(target_position[2])
        msg.velocity.x = float(target_velocity[0])
        msg.velocity.y = float(target_velocity[1])
        msg.velocity.z = float(target_velocity[2])
        msg.acceleration.x = 0.0
        msg.acceleration.y = 0.0
        msg.acceleration.z = 0.0
        msg.yaw = float(target_yaw)
        msg.yaw_dot = float(target_yaw_rate)
        self.pos_cmd_pub.publish(msg)

    def timer_callback(self):
        if self.current_position is None:
            if not self.pose_wait_log_sent:
                self.pose_wait_log_sent = True
                self.get_logger().warn('Waiting for PX4 pose before forwarding cmd_vel')
            return

        if self.pose_wait_log_sent:
            self.pose_wait_log_sent = False
            self.get_logger().info('Received PX4 pose, cmd_vel bridge is active')

        has_recent_cmd = self.has_recent_cmd_vel()
        if not has_recent_cmd:
            return

        target_vel_x, target_vel_y = self.body_to_world_planar_velocity(
            self.latest_cmd_linear_x,
            self.latest_cmd_linear_y,
        )

        target_position = [
            self.current_position[0] + target_vel_x * self.position_lookahead_sec,
            self.current_position[1] + target_vel_y * self.position_lookahead_sec,
            self.current_position[2] + self.latest_cmd_linear_z * self.position_lookahead_sec,
        ]
        target_yaw = self.wrap_angle(self.current_yaw + self.latest_cmd_yaw_rate * self.yaw_lookahead_sec)
        target_velocity = [target_vel_x, target_vel_y, self.latest_cmd_linear_z]
        self.publish_pos_cmd(
            target_position=target_position,
            target_velocity=target_velocity,
            target_yaw=target_yaw,
            target_yaw_rate=self.latest_cmd_yaw_rate,
        )


def main(args=None):
    rclpy.init(args=args)
    node = CmdVelToPosCmdBridge()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass

    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()
